packages/client/lib/miner/miner.ts

```diff
@@ -0,0 +1,206 @@
+  }
+
+  /**
+   * Sets the timeout for the next block assembly
+   */
+  private queueNextAssembly(timeout?: number) {
+    this._nextAssemblyTimeoutId && clearTimeout(this._nextAssemblyTimeoutId)
+    this._nextAssemblyTimeoutId = setTimeout(this.assembleBlock.bind(this), timeout ?? this.period)
+  }
+
+  /**
+   * Sets the next block assembly to latestBlock.timestamp + period
+   */
+  private async chainUpdated() {
+    const latestBlock = await this.latestBlock()
+    const target = latestBlock.header.timestamp.muln(1000).addn(this.period).sub(new BN(Date.now()))
+    const timeout = BN.max(new BN(0), target).toNumber()
+    this.config.logger.debug(
+      `Miner: Chain updated with block ${latestBlock.header.number.toNumber()}. Queuing next block assembly in ${Math.round(
+        timeout / 1000
+      )}s`
+    )
+    this.queueNextAssembly(timeout)
+  }
+
+  /**
+   * Start miner
+   */
+  start(): boolean {
+    if (!this.config.mine || this.running) {
+      return false
+    }
+    this.running = true
+    if (this.config.chainCommon.consensusType() === ConsensusType.ProofOfAuthority) {
+      this.period = this.config.chainCommon.consensusConfig().period * 1000
+    }
+    this.config.events.on(Event.CHAIN_UPDATED, this.chainUpdated.bind(this))
+    this.queueNextAssembly()
+    this.config.logger.info(`Miner started. Assembling next block in ${this.period / 1000}s`)
+    return true
+  }
+
+  /**
+   * Assembles a block from txs in the TxPool and adds it to the chain.
+   * If a new block is received while assembling it will abort.
+   */
+  async assembleBlock() {
+    if (this.assembling) {
+      return
+    }
+    this.assembling = true
+
+    // Abort if a new block is received while assembling this block
+    let interrupt = false
+    const setInterrupt = () => {
+      interrupt = true
+    }
+    this.config.events.on(Event.CHAIN_UPDATED, setInterrupt)
+
+    const parentBlock = await this.latestBlock()
+    const { gasLimit } = parentBlock.header
+    const [signerAddress, signerPrivKey] = this.config.accounts[0]
+
+    // Use a copy of the vm to not modify the existing state.
+    // The state will be updated when the newly assembled block
+    // is inserted into the canonical chain.
+    const vmCopy = this.synchronizer.execution.vm.copy()
+
+    // If we are on the genesis block the canonical
+    // genesis state has not been initialiezd yet.
+    if (parentBlock.header.number.isZero()) {
+      await vmCopy.stateManager.generateCanonicalGenesis()
+    }
+
+    let difficulty
+    if (this.config.chainCommon.consensusType() === ConsensusType.ProofOfAuthority) {
+      // Determine if signer is INTURN (2) or NOTURN (1)
+      const signers = vmCopy.blockchain.cliqueActiveSigners()
+      const signerIndex = signers.findIndex((address) => address.equals(signerAddress))
+      const inTurn = parentBlock.header.number.addn(1).modn(signers.length) === signerIndex
+      difficulty = inTurn ? 2 : 1
+    }
```

Comment:
I guess this whole process - 1. getting the tip of the chain block, 2. getting the active signers and 3. then determining if a signer is in turn or not would be a good candidate for a utility method on the blockchain package, something like `Blockchain.cliqueSignerInTurn(signer)` (or similar).

Reply:
yeah, sure, making a util method of this on `blockchain` sounds like a great idea.

I already had this code from `BlockHeader.validateCliqueDifficulty(blockchain)` so I can also dedupe it there when added 👍